### PR TITLE
drivers: memc: rework FlexSPI memc API to better support multiple flash devices

### PR DESCRIPTION
**Pull request by danieldegrasse**
Rework the FlexSPI MEMC API in order to better support multiple flash devices. This change requires that the flash chip's port and LUT table be configured in one API call, so that the memc driver can dynamically select the LUT offset to use for a given port

The advantage to this approach is that multiple devices on a FlexSPI bus no longer need to dynamically set their LUT sequence indices when setting up transfers. Instead, all devices can initiate transfers as though their LUT starts at offset 0, and the memc driver will select the correct LUT offset for the transfer based on the port.

This change has been verified on the following boards:

- RT1060: validated with MCUBoot (flash_mcux_flexspi_nor.c)
- RT1050: validated with MCUBoot (flash_mcux_flexspi_hyperflash.)
- RT685: validated with MCUBoot (flash_mcux_flexspi_mx25um51345g.c)
- RT595: validated with memc sample (memc_mcux_flexspi_aps6408l.c)

### CODE REVIEW

**Comment by danieldegrasse** [comment]
FYI @pdgendt and @moradelk- this change will affect the PSRAM drivers you all contributed

**Review by pdgendt — DISMISSED** [review]

File: samples/drivers/memc/src/main.c [hunk 1]
```diff
@@ -50,7 +53,7 @@ int main(void)
 		memc_write_buffer[i] = (uint8_t)i;
 	}
 	printk("Writing to memory region with base 0x%0x, size 0x%0x\n\n",
-		MEMC_BASE, MEMC_SIZE);
+		((uint32_t) MEMC_BASE), MEMC_SIZE);
```
Comment (de-nordic) [inline samples/drivers/memc/src/main.c:56]: This truncates the MEMC_BASE or promotes it up? Because in first case it may give incorrect info.

**Comment by MaureenHelm** [comment]
@pdgendt @de-nordic please take a look

File: drivers/memc/memc_mcux_flexspi.c [hunk 2]
```diff
@@ -108,20 +107,65 @@ int memc_flexspi_update_clock(const struct device *dev,
...
+	data->port_luts[port].lut_used = lut_count;
+	tmp_config.ARDSeqIndex += data->port_luts[port].lut_offset;
+	tmp_config.AWRSeqIndex += data->port_luts[port].lut_offset;
+
+	/* Lock IRQs before reconfiguring FlexSPI, to prevent XIP */
+	key = irq_lock();
+
+	FLEXSPI_SetFlashConfig(data->base, &tmp_config, port);
+	FLEXSPI_UpdateLUT(data->base, data->port_luts[port].lut_offset,
+				lut_ptr, lut_count);
```
Comment (pdgendt) [inline drivers/memc/memc_mcux_flexspi.c:167]: Nit: indentation

**Review by pdgendt — DISMISSED** [review]

File: drivers/memc/memc_mcux_flexspi.c (hunk 2 above)
Comment (danieldegrasse) [reply drivers/memc/memc_mcux_flexspi.c:167]: Fixed

**Review by pdgendt — APPROVED** [review]

**Review by MaureenHelm — APPROVED** [review]

**Review by de-nordic — APPROVED** [review]

**Comment by fabiobaltieri** [comment]
@danieldegrasse should we hold this for 3.6 or should this count as a fix?

**Comment by danieldegrasse** [comment]
> @danieldegrasse should we hold this for 3.6 or should this count as a fix?

Personally, I think the rework is too large to merge during the code freeze, especially since this is simply preparation for the fix in #61905

**Pull request merged**